libbeat/processors/communityid/communityid.go

```diff
@@ -0,0 +1,265 @@
+
+func init() {
+	processors.RegisterPlugin("community_id", New)
+}
+
+type processor struct {
+	config
+	log    *logp.Logger
+	hasher flowhash.Hasher
+}
+
+// New constructs a new processor that computes community ID flowhash. The
+// values that are incorporated into the hash vary by protocol.
+//
+// TCP / UDP / SCTP:
+//   IP src / IP dst / IP proto / source port / dest port
+//
+// ICMPv4 / ICMPv6:
+//   IP src / IP dst / IP proto / ICMP type + "counter-type" or code
+//
+// Other IP-borne protocols:
+//   IP src / IP dst / IP proto
+func New(cfg *common.Config) (processors.Processor, error) {
+	c := defaultConfig()
+	if err := cfg.Unpack(&c); err != nil {
+		return nil, errors.Wrap(err, "fail to unpack the community_id configuration")
+	}
+
+	return newFromConfig(c)
+}
+
+func newFromConfig(c config) (*processor, error) {
+	return &processor{
+		config: c,
+		log:    logp.NewLogger(logName),
+		hasher: flowhash.CommunityID,
+	}, nil
+}
+
+func (p *processor) String() string {
+	return fmt.Sprintf("community_id=[target=%s, fields=["+
+		"source_ip=%v, source_port=%v, "+
+		"destination_ip=%v, destination_port=%v, "+
+		"transport_protocol=%v, "+
+		"icmp_type=%v, icmp_code=%v]",
+		p.Target, p.Fields.SourceIP, p.Fields.SourcePort,
+		p.Fields.DestinationIP, p.Fields.DestinationPort,
+		p.Fields.TransportProtocol, p.Fields.ICMPType, p.Fields.ICMPCode)
+}
+
+func (p *processor) Run(event *beat.Event) (*beat.Event, error) {
+	// If already set then bail out.
+	_, err := event.GetValue(p.Target)
+	if err == nil {
+		return event, nil
+	}
+
+	flow := p.buildFlow(event)
+	if flow == nil {
+		return event, nil
+	}
+
+	id := p.hasher.Hash(*flow)
+	_, err = event.PutValue(p.Target, id)
+	return event, err
+}
+
+func (p *processor) buildFlow(event *beat.Event) *flowhash.Flow {
+	var flow flowhash.Flow
+
+	// source ip
+	v, err := event.GetValue(p.Fields.SourceIP)
+	if err != nil {
+		return nil
+	}
+	var ok bool
+	flow.SourceIP, ok = tryToIP(v)
+	if !ok {
+		return nil
+	}
+
+	// destination ip
+	v, err = event.GetValue(p.Fields.DestinationIP)
+	if err != nil {
+		return nil
+	}
+	flow.DestinationIP, ok = tryToIP(v)
+	if !ok {
+		return nil
+	}
+
+	// protocol
+	v, err = event.GetValue(p.Fields.TransportProtocol)
+	if err != nil {
+		return nil
+	}
+	flow.Protocol, ok = tryToIANATransportProtocol(v)
+	if !ok {
+		return nil
+	}
+
+	switch flow.Protocol {
+	case tcpProtocol, udpProtocol, sctpProtocol:
+		// source port
+		v, err = event.GetValue(p.Fields.SourcePort)
+		if err != nil {
+			return nil
+		}
+		flow.SourcePort, ok = tryToUint16(v)
+		if !ok || flow.SourcePort == 0 {
+			return nil
+		}
+
+		// destination port
+		v, err = event.GetValue(p.Fields.DestinationPort)
+		if err != nil {
+			return nil
+		}
+		flow.DestinationPort, ok = tryToUint16(v)
+		if !ok || flow.DestinationPort == 0 {
+			return nil
+		}
+	case icmpProtocol, icmpIPv6Protocol:
+		v, err = event.GetValue(p.Fields.ICMPType)
+		if err != nil {
+			return nil
```

Comment:
I'm wondering if for ICMP(v6) it's better to ignore when the type and/or code fields are missing and still generate the hash with these fields zeroed, as in some cases they might not be available.

Reply:
I made the change, can you please take a look and see if this is what you meant. I went with an "all or nothing" approach. So if you have both code and type fields then they are incorporated into the hash, otherwise zeros are used. Maybe I'm being too strict.

Reply:
:+1: your approach makes more sense than my suggestion. Using one field without the other is useless.